### ISSUE TITLE
Training a GPT-2 from scratch in Greek-text, results in a low perplexity score of 7 after 15 epochs. Is it normal that score?

### ISSUE THREAD

**Issue by niklaras** [huggingface/transformers#5391]
# ❓ Questions & Help

## Details

**A link to original question on Stack Overflow**:

**Comment by niklaras** [comment]
My data for train is about 4.6Gb and is constructed as 5 sentences per line . The data for the evaluation is about 450Mb constructed with the same way.

**Comment by xx-zhou16** [comment]
I think it is normal.
For perplexity, the lower, the better.
By the way, can you share your script for preprocessing data?

**Comment by niklaras** [comment]
@xx-zhou16 Ok i found a mistake i had...when i was computing the loss i wasn't ignoring the pad_token . Now i train it again , i think now the perplexity score will stop about 11, which is again low. 
I use the line by LineByLineTextDataset :

class LineByLineTextDataset(Dataset):
    def __init__(self, tokenizer: PreTrainedTokenizer, args, file_path: str, block_size=512):
        assert os.path.isfile(file_path)
        # Here, we do not cache the features, operating under the assumption
        # that we will soon use fast multithreaded tokenizers from the
        # `tokenizers` repo everywhere =)
        logger.info("Creating features from dataset file at %s", file_path)

        with open(file_path, encoding="utf-8") as f:
            lines = [line for line in f.read().splitlines() if (len(line) > 0 and not line.isspace())]
        self.examples = tokenizer.batch_encode_plus(lines, add_special_tokens=True, truncation=True, max_length=block_size)["input_ids"]
        
    def __len__(self):
        return len(self.examples)

    def __getitem__(self, i):
        return torch.tensor(self.examples[i], dtype=torch.long)


def load_and_cache_examples(args, tokenizer, evaluate=False):
    file_path = args.eval_data_file if evaluate else args.train_data_file
    if args.line_by_line:
        return LineByLineTextDataset(tokenizer, args, file_path=file_path, block_size=args.block_size)

## padding and creating an attention mask 
def collate(examples: List[torch.Tensor]):
        padding_value = 0 if tokenizer._pad_token is None else tokenizer.pad_token_id
        input_ids = pad_sequence(examples, batch_first=True, padding_value=padding_value)

        max_length = input_ids.shape[1]
        attention_mask = torch.stack([torch.cat([torch.ones(len(t), dtype=torch.long), torch.zeros(max_length - len(t), dtype=torch.long)]) for t in examples])

        return input_ids, attention_mask


and that is in the evaluation , for the perplexity score computation.

    for batch in tqdm(eval_dataloader, desc="Evaluating"):
        input_ids, attention_mask = batch
        inputs, labels = (input_ids, input_ids)

        inputs = inputs.to(args.device)
        labels = labels.to(args.device)
        attention_mask = attention_mask.to(args.device)

        with torch.no_grad():
            outputs = model(inputs, labels=labels, attention_mask=attention_mask)
            lm_loss = outputs[0].mean().item()
            eval_loss += lm_loss
        nb_eval_steps += 1


    eval_loss = eval_loss / nb_eval_steps
    perplexity = torch.exp(torch.tensor(eval_loss))
